markdowns/19_Text_to_Image_search_pipeline_with_MultiModal_Retriever.md

```diff
@@ -0,0 +1,190 @@
+    for filename in os.listdir(f"./{doc_dir}/spirit-animals/")
+]
+
+document_store.write_documents(images)
+```
+
+# Initializing the Retriever
+
+Retrievers sift through all the images and return only those that are relevant based on the input query. Here we are using the [OpenAI CLIP model](https://github.com/openai/CLIP/blob/main/model-card.md) to embed images. For more details on supported modalities, see [MultiModalRetriever](https://docs.haystack.deepset.ai/docs/retriever#multimodal-retrieval).
+
+
+```python
+from haystack.nodes.retriever.multimodal import MultiModalRetriever
+
+retriever_text_to_image = MultiModalRetriever(
+    document_store=document_store,
+    query_embedding_model = "sentence-transformers/clip-ViT-B-32",
+    query_type="text",
+    document_embedding_models = {"image": "sentence-transformers/clip-ViT-B-32"}
+)
+
+document_store.update_embeddings(retriever=retriever_text_to_image)
+```
+
+# Creating the MultiModal search Pipeline
+
+We are populating a Pipeline with a MultiModalRetriever node. This search pipeline queries the image database with text and returns the most relevant images.
+
+
+```python
+from haystack import Pipeline
+
+pipeline = Pipeline()
+pipeline.add_node(
+    component=retriever_text_to_image, 
+    name="retriever_text_to_image", 
+    inputs=["Query"]
+)
+```
+
+# Searching Through the Images
+
+Use the pipeline `run()` method to query the images in the DocumentStore. The query argument is where you type your text query. Additionally, you can set the number of images you want the MultiModalRetriever to return using the `top-k` parameter. To learn more about setting arguments, see [Pipeline Arguments](https://docs.haystack.deepset.ai/docs/pipelines#arguments).
+
+
+```python
+results = pipeline.run(
+    query="Animal who lives in the water",
+    params={"retriever_text_to_image": {"top_k": 3}}
+)
+
+results = sorted(results["documents"], key=lambda d: d.score, reverse=True)
+
+for doc in results:
+    print(doc.score, doc.content)
+```
+
+Here are some more query strings you could try out:
+
+1.   King of the Jungle
+2.   Fastest animal
+3.   Bird who can see clearly even in the dark
+
+
+
+You can also vizualize these images with their score easily with below code.
+
+
+
+
+
+```python
+from io import BytesIO
+from PIL import Image, ImageDraw, ImageOps
+from IPython.display import display, Image as IPImage
+
+def display_img_array(ima, score):
+    im = Image.open(ima)
+    img_with_border = ImageOps.expand(im ,border=20, fill='white')
+
+    # Add Text to an image
+    img = ImageDraw.Draw(img_with_border)
+    img.text((20, 0), f"Score: {score},    Path: {ima}", fill=(0, 0, 0))
+
+    bio = BytesIO()
+    img_with_border.save(bio, format='png')
+    display(IPImage(bio.getvalue(), format='png'))
+
+images_array = [doc.content for doc in results]
+scores = [doc.score for doc in results]
+for ima, score in zip(images_array, scores):
+    display_img_array(ima, score)
+```
+
```

Comment:
How about adding a result here: Congratulations! You've created a search system that returns images of animals in answer to a text query.

Reply:
Added